astrovascpy/PetscBinaryIO.py

```diff
@@ -65,7 +65,7 @@ def get_conf():
         except importlib.metadata.PackageNotFoundError:
             try:
                 petsc4py = importlib.metadata.distribution("petsc4py")
-                if "conda" in str(petsc4py.files[0].locate()):
+                if "conda" or "tox" in str(petsc4py.files[0].locate()):
```

Comment:
@tristan0x This is the line that breaks the `test_petsc_installation`. Because in the tox environment, there is no "conda" in the path.

Reply:
But according to `setup.sh`, petsc4py is not supposed to be installed inside tox.

Reply:
So if you can give a **full** description of the error you experience, that would be helpful. It comprises the list of commands you ran, their outputs if errors occurred, the list of packages in your conda environment, if you use setup.sh or not...

Reply:
So, since we decided to make the repository public, I expect that a user could use `tox` on his local machine. 
But at the moment, `tox` is not working. Neither locally nor on BB5.  And all problems come from `petsc4py`, which cannot be installed with `pip`, but only with `conda`.

Many functions that we are testing use `petsc`, and `petsc4py` must be available within the `tox` virtual environment.

What I did is the following:
- Starting from the main branch version on my local computer. I created the virtual environment using `setup.sh`. 
- Then `conda install tox` and `conda install tox-conda`.
- Then I modified the `tox.ini` as it is now in this branch, to allow for the use of `tox-conda` and to make it install `petsc4py` inside the `venv`.
- Then running the tests with: `tox -re py310` I get the following error:

```
    def test_petsc_installation():
        precision, indices, complex = get_conf()
>       assert (
            complex is False
        ), "PETSc needs to be compiled with configure option --with-scalar-type=real"
E       AssertionError: PETSc needs to be compiled with configure option --with-scalar-type=real
E       assert None is False
```

In this PR the error already is fixed.

Reply:
I was able to use tox successfully (I only have the test_solve_linear error on ARM) on Apple M1, a Linux workstation, and BB5. The fix for the error you experience must reside in the environment from which you launch tox, not tox itself.

tox is the standard way to test Python packages in BlueBrain through the very nice GitLab CI/CD library that nse maintains for us. tox provides a straightforward way to reproduce exactly what the CI does. tox is agnostic to the Python environment where it runs (system Python, conda, mamba, spack, Nix, ...), and we should keep it that way. 
 Conda is one of the dozen kinds of Python environments that can be used, I don't see why it should be used specifically on all platforms.

Besides, AstroVasPy will be delivered by spack shortly, and provided like the rest of the BlueBrain code through environment modules on BB5. So conda does not really fit in the picture.

> but importing modules inside a virtual environment doesn't seem a great idea.
IMO there is no problem coupling environment modules with Python virtual environments as long as they share the same Python interpreter, which is the case on BB5. Could you explain your thoughts behind your comment? Have you experienced issues related to the mixin on both?

Reply:
Sorry, on BB5 it works. Only on my local machine (an Apple M2) I have issues. 
This is what I did:
- cloned the repo.  And using the `main` branch
- run `source setup.sh`
- run `pip install tox` (or `conda install tox`)
- run `tox`.
```
E   ModuleNotFoundError: No module named 'mpi4py'
```
Actually I was expecting this error, since mpi4py and petsc4py are not included in the dependencies in `setup.py`. (in BB5 is it getting them from the modules? )

So, what did you do differently?  Did I miss something?
If I did something wrong at this step, there is no need to read the following.

------------------------------------------------------------------------------------------------------------

Regarding the error in `test_petsc_installation`.
(here I am using the `conda-tox` plugin that I introduced in this branch)

When the tests runs inside the tox environment `/AstroVascPy/.tox/py310`, the code in line 68 `str(petsc4py.files[0].locate())` is equal to 
```
'/AstroVascPy/.tox/py310/lib/python3.10/site-packages/petsc4py-3.20.0.dist-info/INSTALLER'
```
while running the test in the conda environment `bfs_env` the code has value
```
'/miniconda3/envs/bfs_env/lib/python3.10/site-packages/petsc4py-3.17.3.dist-info/INSTALLER'
```
This is the reason of the fix  in line 68.
(notice also the 2 different versions of petsc4py, which is another unwanted thing)

----------------------------------------------------------------------------------------------------------------

Regarding tox, I don’t have any problem with using tox for the CI.  My comment was just to mention that since the modules can change in the future, we can fail to reproduce the same tox environment.